atom/src/catom.cpp

```diff
@@ -370,6 +372,98 @@ CAtom_sizeof( CAtom* self, PyObject* args )
+    cppy::ptr stateptr = PyDict_New();
+    if ( !stateptr )
+        return PyErr_NoMemory();  // LCOV_EXCL_LINE
+
+    cppy::ptr selfptr(pyobject_cast(self), true);
+
+    // Copy __dict__ if present. Using hasattr / getattr makes it slower
+    // than the py version hence the _PyObject_GetDictPtr.
+    if ( PyObject** dict = _PyObject_GetDictPtr( selfptr.get() ) )
+    {
+        if ( PyDict_Update( stateptr.get(), dict[0] ) )
+            return 0;
+    }
+
+    // Copy __slots__ if present. This assumes copyreg._slotnames was called
+    // during AtomMeta's initialization
+    {
+        cppy::ptr typeptr = PyObject_Type(selfptr.get());
+        if (!typeptr)
+            return 0;
+        cppy::ptr slotnamesptr = typeptr.getattr("__slotnames__");
+        if (!slotnamesptr.get())
+            return 0;
+        if (!PyList_CheckExact(slotnamesptr.get()))
+            return cppy::system_error( "slot names" );
+        for ( Py_ssize_t i=0; i < PyList_GET_SIZE(slotnamesptr.get()); i++ )
+        {
+            PyObject *name = PyList_GET_ITEM(slotnamesptr.get(), i);
+            cppy::ptr value = selfptr.getattr(name);
+            if (!value || PyDict_SetItem(stateptr.get(), name, value.get()) )
+                return  0;
+        }
+    }
+
+    cppy::ptr membersptr = selfptr.getattr(atom_members);
+    if ( !membersptr || !PyDict_CheckExact( membersptr.get() ) )
+        return cppy::system_error( "atom members" );
+
+    PyObject *name, *member;
+    Py_ssize_t pos = 0;
+    while ( PyDict_Next(membersptr.get(), &pos, &name, &member) ) {
+        switch (member_cast(member)->get_setattr_mode()) {
+            case SetAttr::Mode::Constant:
+            case SetAttr::Mode::ReadOnly:
+            case SetAttr::Mode::Event:
+            case SetAttr::Mode::Signal:
+            case SetAttr::Mode::Property:
+                continue;
+            default:
+                break;
+        }
+        cppy::ptr value = member_cast(member)->getattr(self);
+        if (!value || PyDict_SetItem(stateptr.get(), name, value.get()) )
+            return  0;
+    }
+
+    // Frozen state
+    if ( self->is_frozen() && PyDict_SetItem(stateptr.get(), atom_flags, frozen_bit) )
+        return 0;
+
+    return stateptr.release();
+}
+
+PyObject*
+CAtom_setstate( CAtom* self, PyObject* args )
+{
+    if( PyTuple_GET_SIZE( args ) != 1 )
+        return cppy::type_error( "__setstate__() takes exactly one argument" );
+    PyObject* state = PyTuple_GET_ITEM( args, 0 );
+    cppy::ptr itemsptr = PyMapping_Items(state);
+    if ( !itemsptr )
+        return 0;
+    cppy::ptr selfptr(pyobject_cast(self), true);
+    bool freeze = false;
+    for ( Py_ssize_t i = 0; i < PyMapping_Size(state); i++ ) {
+        PyObject* item = PyList_GET_ITEM(itemsptr.get(), i);
+        PyObject* key = PyTuple_GET_ITEM(item , 0);
+        PyObject* value = PyTuple_GET_ITEM(item , 1);
+        if ( PyObject_RichCompareBool(key, atom_flags, Py_EQ) == 1 && PyLong_Check(value) )
```

Comment:
I still believe there could be value in making the frozen key always present and to pop it early on.

Reply:
There didn't seem to be any measurable performance impact when adding the extra compare. Maybe it will if theres a lot of fields, idk.

Also setstate can be used outside of pickle (my use case does not involve pickle at all) so there's no guarantee that the flag will be there.  In my case state is not a dict but a row from sqlalchemy that implements a mapping.

Reply:
Fair enough for your use case but could you add a comment about it then ?

Reply:
It now just checks if the flag is there (and deletes it)  which should work fine for my use case since that will always be skipped.

Do you think it should put the flag back in so the state is left unchanged?

Reply:
So I tried to look up what invariant setstate is expected to maintain but did not find much. Basically it seems from the docs that when a class defines a custom getstate it can expect to get the getstate output as input of setstate.  

So assuming mutability sounds reasonable but this does not really align with your use case. Could you maybe point me to the origin of the behavior you are relying on ?